### PR TITLE
Celery autoscaler

### PR DESCRIPTION
**Pull request by squeakymouse**
# Pull Request Summary

Move over the Celery autoscaler

## Test Plan and Usage Guide

Installing a test deployment autoscaled an async endpoint from 0 to 1 pod

### CODE REVIEW

File: charts/model-engine/templates/celery_autoscaler_stateful_set.yaml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        node-lifecycle: normal
+      tolerations:
+        - key: CriticalAddonsOnly
+          operator: Equal
+          value: 'true'
+          effect: NoSchedule
+      serviceAccountName: ml-k8s-admin
+      volumes:
+      - configMap:
+          name: ml-worker-config
```
Comment (yunfeng-scale) [inline charts/model-engine/templates/celery_autoscaler_stateful_set.yaml:93]: `{{ .Values.aws.configMap.name }}` ?

File: charts/model-engine/templates/celery_autoscaler_stateful_set.yaml [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        - name: NUM_SHARDS
+          value: '{{ $num_shards }}'
+        image: "{{ $.Values.image.gatewayRepository }}:{{ $tag }}"
+        imagePullPolicy: Always
+        name: main
+        resources:
+          requests:
+            cpu: 1000m
+        volumeMounts:
+        - mountPath: /root/.aws/config
```
Comment (saiatmakuri) [inline charts/model-engine/templates/celery_autoscaler_stateful_set.yaml:80]: /opt?

File: model-engine/model_engine_server/core/celery_autoscaler.py [hunk 3]
```diff
@@ -0,0 +1,635 @@
...
+
+
+ELASTICACHE_REDIS_BROKER = "redis-elasticache-message-broker-master"
+SQS_BROKER = "sqs-message-broker-master"
+
+UPDATE_DEPLOYMENT_MAX_RETRIES = 10
+
+SQS_SAMPLE_COUNT = 10
+
+logger = make_logger("model_engine_server.core.celery_autoscaler")
```
Comment (saiatmakuri) [inline model-engine/model_engine_server/core/celery_autoscaler.py:51]: nit: use `logger_name()` that you can import as well

File: model-engine/model_engine_server/core/celery_autoscaler.py [hunk 4]
```diff
@@ -0,0 +1,635 @@
...
+
+
+def excluded_namespaces():
+    try:
+        from plugins.celery_autoscaler_dependencies import CELERY_AUTOSCALER_EXCLUDED_NAMESPACES
+
+        return CELERY_AUTOSCALER_EXCLUDED_NAMESPACES
+    except ModuleNotFoundError:
+        return []
+    finally:
```
Comment (saiatmakuri) [inline model-engine/model_engine_server/core/celery_autoscaler.py:40]: nit: unnecessary

File: model-engine/model_engine_server/core/celery_autoscaler.py [hunk 5]
```diff
@@ -0,0 +1,635 @@
...
+            for db_index in get_all_db_indexes()
+        }
+    elif broker_type == "sqs":
+        # for sqs we will get active/reserved counts directly from sqs as opposed to using
+        #   an inspect object
+        inspect = {}
+    else:
+        raise ValueError("broker_type not redis or sqs, how did we get here?")
+
+    env = os.getenv("DD_ENV", "training")
```
Comment (saiatmakuri) [inline model-engine/model_engine_server/core/celery_autoscaler.py:561]: do we want to default to training here?

File: model-engine/model_engine_server/core/celery_autoscaler.py [hunk 6]
```diff
@@ -0,0 +1,635 @@
+import asyncio as aio
```
Comment (saiatmakuri) [inline model-engine/model_engine_server/core/celery_autoscaler.py:1]: could this be moved to the `celery/` folder?

File: model-engine/model_engine_server/core/celery/celery_autoscaler.py [hunk 7]
```diff
@@ -0,0 +1,633 @@
...
+
+            if deployment_broker != autoscaler_broker:
+                logger.debug(
+                    f"Skipping deployment {deployment_name}; deployment's broker {deployment_broker} is not {autoscaler_broker}"
+                )
+                continue
+
+            for f in dataclasses.fields(CeleryAutoscalerParams):
+                k = f.name
+                # TODO: this technically can remain the same, since Celery deployments won't have the `channel` field and
```
Comment (seanshi-scale) [inline model-engine/model_engine_server/core/celery/celery_autoscaler.py:104]: I guess we can remove this comment probably? there is no NATS autoscaler anymore

File: charts/model-engine/templates/celery_autoscaler_stateful_set.yaml [hunk 8]
```diff
@@ -0,0 +1,98 @@
+{{- if .Values.celery_autoscaler.enabled }}
+{{- $env := .Values.context }}
+{{- $tag := .Values.tag }}
+{{- $service_identifier := .Values.serviceIdentifier }}
+{{- $num_shards := ternary 30 3 (eq .Values.context "prod") }}
+{{- range $base_app := tuple "celery-autoscaler-elasticache" "celery-autoscaler-sqs" }}
```
Comment (seanshi-scale) [inline charts/model-engine/templates/celery_autoscaler_stateful_set.yaml:6]: it's probably fine to have this range but I think that we'd be able to choose which one based on whether we're having endpoints use redis or sqs at some point

File: charts/model-engine/templates/celery_autoscaler_stateful_set.yaml [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+{{- $app := ternary (printf "%s-%s" $base_app $service_identifier) $base_app (not (empty $service_identifier)) }}
+{{- $broker_name := ternary "redis-elasticache-message-broker-master" "sqs-message-broker-master" (eq $base_app "celery-autoscaler-elasticache") }}
+---
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  labels:
+    app: {{ $app }}
+    team: infra
```
Comment (yixu34) [inline charts/model-engine/templates/celery_autoscaler_stateful_set.yaml:15]: These tags shouldn't be hardcoded, they're internal.

**Review by yixu34 — COMMENTED** [review]
Let's parameterize the cost tags.

File: charts/model-engine/templates/celery_autoscaler_stateful_set.yaml (hunk 9 above)
Comment (yunfeng-scale) [reply charts/model-engine/templates/celery_autoscaler_stateful_set.yaml:15]: see other deployments. something like
```
    {{- include "modelEngine.selectorLabels.celeryAutoScaler" . | nindent 4 }}
    {{- include "modelEngine.labels" . | nindent 4 }}
    tags.datadoghq.com/service: {{ include "modelEngine.celeryAutoScalerName" . }}
```
should

File: charts/model-engine/templates/celery_autoscaler_stateful_set.yaml [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+  selector:
+    matchLabels:
+      app: {{ $app }}
+  template:
+    metadata:
+      annotations:
+        ad.datadoghq.com/main.logs: '[{"service": "{{ $app }}", "source": "python"}]'
+        sidecar.istio.io/inject: "false"
+      labels:
+        app: {{ $app }}
```
Comment (yunfeng-scale) [inline charts/model-engine/templates/celery_autoscaler_stateful_set.yaml:33]: these tags should also be templatized

File: charts/model-engine/templates/celery_autoscaler_stateful_set.yaml (hunk 9 above)
Comment (yunfeng-scale) [reply charts/model-engine/templates/celery_autoscaler_stateful_set.yaml:15]: yeah i think that file should parametrized tags. don't see why balloon deployment needs a different `product` tag

File: charts/model-engine/values_sample.yaml [hunk 11]
```diff
@@ -91,6 +91,14 @@ autoscaling:
   prewarming:
     enabled: false
 
+# for async endpoints, Celery autoscaler scales the number of pods based on number of requests
+# message_broker should be either "sqs" or "elasticache"
+# num_shards is number of instances of the autoscaler
+celery_autoscaler:
+  enabled: true
+  message_broker: sqs
```
Comment (seanshi-scale) [inline charts/model-engine/values_sample.yaml:99]: this is already defined as the root-level "messageBrokerType" (can see that at the bottom of this file)

**Review by seanshi-scale — COMMENTED** [review]
had one nit about deduplicating info in the helm values but the rest looks good

**Review by yunfeng-scale — APPROVED** [review]

**Pull request merged**